kernel/src/main/scala/org/apache/toree/boot/layer/ComponentInitialization.scala

```diff
@@ -36,22 +37,39 @@ import org.apache.toree.utils.{LogLike, FileUtils}
 import scala.collection.JavaConverters._
 import org.apache.toree.plugins.AllInterpretersReady
 
+import scala.concurrent.Future
+import scala.concurrent.ExecutionContext.Implicits.global
+
 /**
  * Represents the component initialization. All component-related pieces of the
  * kernel (non-actors) should be created here. Limited items should be exposed.
  */
 trait ComponentInitialization {
+
+  /**
+   * Start initializing slow components, specifically the scala interpreter. These are broken out so we
+   * can start them as early as possible.
+   *
+   * @param config the config used for initialization
+   */
+  def initializeSlowComponents(config: Config): (InterpreterManager, Option[Future[ScalaInterpreter]])
+
   /**
    * Initializes and registers all components (not needed by bare init).
    *
    * @param config The config used for initialization
    * @param actorLoader The actor loader to use for some initialization
+   * @param interpreterManager the interpreterManager
+   * @param maybeEventualScalaInterp  a future that will hold the Scala Interpreter when its ready
    */
   def initializeComponents(
-    config: Config, actorLoader: ActorLoader
+      config: Config,
+      actorLoader: ActorLoader,
+      interpreterManager: InterpreterManager,
+      maybeEventualScalaInterp: Option[Future[ScalaInterpreter]]
   ): (CommStorage, CommRegistrar, CommManager, Interpreter,
     Kernel, DependencyDownloader, MagicManager, PluginManager,
```

Comment:
I know this is not your code, but this is a LOT of stuff to return from one method. Might be better organized as member variables in one case class.

Reply:
agreed